security-csrf/src/main/java/io/micronaut/security/csrf/filter/CsrfFilter.java

```diff
@@ -0,0 +1,194 @@
+import io.micronaut.security.authentication.AuthorizationException;
+import io.micronaut.security.csrf.resolver.CsrfTokenResolver;
+import io.micronaut.security.csrf.resolver.ReactiveCsrfTokenResolver;
+import io.micronaut.security.csrf.validator.CsrfTokenValidator;
+import io.micronaut.security.filters.SecurityFilter;
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * {@link RequestFilter} which validates CSRF tokens and rejects a request if the token is invalid.
+ * Which requests are intercepted can be controlled via {@link io.micronaut.security.csrf.CsrfConfiguration}.
+ * @author Sergio del Amo
+ * @since 4.11.0
+ */
+@Internal
+@Requires(property = CsrfFilterConfigurationProperties.PREFIX + ".enabled", value = StringUtils.TRUE, defaultValue = StringUtils.TRUE)
+@Requires(classes = { ExceptionHandler.class, HttpRequest.class })
+@Requires(beans = { CsrfTokenValidator.class })
+@ServerFilter(patternStyle = FilterPatternStyle.REGEX,
+        value = "${" + CsrfFilterConfigurationProperties.PREFIX + ".regex-pattern:" + CsrfFilterConfigurationProperties.DEFAULT_REGEX_PATTERN + "}")
+final class CsrfFilter implements Ordered {
+    private static final Logger LOG = LoggerFactory.getLogger(CsrfFilter.class);
+    private final List<ReactiveCsrfTokenResolver<HttpRequest<?>>> reactiveCsrfTokenResolvers;
+    private final List<CsrfTokenResolver<HttpRequest<?>>> csrfTokenResolvers;
+    private final CsrfTokenValidator<HttpRequest<?>> csrfTokenValidator;
+    private final ExceptionHandler<AuthorizationException, MutableHttpResponse<?>> exceptionHandler;
+    private final CsrfFilterConfiguration csrfFilterConfiguration;
+
+    CsrfFilter(CsrfFilterConfiguration csrfFilterConfiguration,
+               List<ReactiveCsrfTokenResolver<HttpRequest<?>>> reactiveCsrfTokenResolvers,
+               List<CsrfTokenResolver<HttpRequest<?>>> csrfTokenResolvers,
+               CsrfTokenValidator<HttpRequest<?>> csrfTokenValidator,
+               ExceptionHandler<AuthorizationException, MutableHttpResponse<?>> exceptionHandler) {
+        this.csrfTokenResolvers = csrfTokenResolvers;
+        this.reactiveCsrfTokenResolvers = reactiveCsrfTokenResolvers.isEmpty()
+                ? reactiveCsrfTokenResolvers
+                : ReactiveCsrfTokenResolver.of(csrfTokenResolvers, reactiveCsrfTokenResolvers);
+        this.csrfTokenValidator = csrfTokenValidator;
+        this.exceptionHandler = exceptionHandler;
+        this.csrfFilterConfiguration = csrfFilterConfiguration;
+    }
+
+    @RequestFilter
+    @Nullable
+    public Publisher<Optional<MutableHttpResponse<?>>> csrfFilter(@NonNull HttpRequest<?> request) {
+        if (!shouldTheFilterProcessTheRequestAccordingToTheHttpMethod(request)) {
+            return proceedRequest();
```

Comment:
Actually it gets even better. I think you can do this:

```suggestion
            return null;
```

Reply:
Actually, hmm, not sure. Would be a nice feature if it doesn't work.

Reply:
It does not work to return `null. `

It does not work to return `Mono.empty()`. 

Only, it works to return `Mono.just(Optional.empty())`

That it is the reason I created:
https://github.com/micronaut-projects/micronaut-core/pull/11270

With the current javadoc, if you did not tell me that to proceed with a request I had to use `Publisher<Optional<HttpResponse>>` and return `Mono.just(Optional.empty())` I would not have figure it out.